### PR TITLE
reduce memory usage of Transport

### PR DESCRIPTION
**Pull request by pawl**
While investigating a redis broker memory leak, I noticed the same issue as py-amqp's [#377](https://github.com/celery/py-amqp/pull/377):
```
Top 10 lines
#1: /celery_app/kombu/kombu/transport/virtual/base.py:911: 782.4 KiB
    ARRAY_TYPE_H, range(self.channel_max, 0, -1),
<truncated>
```

The array that's used to store available connection ids is taking up an unnecessary amount of memory (more than anything else while running celery), and it's making memory leaks worse.

This PR reduces the memory usage of the `Transport` by reversing that logic and turning `_avail_channel_ids` into `_used_channel_ids`. This will only store the channel ids that are currently used, which is almost always going to be a much smaller array.

### CODE REVIEW

File: kombu/transport/virtual/base.py [hunk 1]
```diff
@@ -462,9 +462,15 @@ def __init__(self, connection, **kwargs):
             typ: cls(self) for typ, cls in self.exchange_types.items()
         }
 
-        try:
-            self.channel_id = self.connection._avail_channel_ids.pop()
-        except IndexError:
+        # Cast to a set for fast lookups, and keep stored as an array for lower memory usage.
+        used_channel_ids = set(self.connection._used_channel_ids)
```
Comment (pawl) [inline kombu/transport/virtual/base.py:466]: Maybe we should put this `_used_channel_ids` stuff into its own method like it is in py-amqp?: https://github.com/celery/py-amqp/blob/6433be13199938f47bbc2826e6585e746baf4f9b/amqp/connection.py#L484-L492

Comment (auvipy) [reply kombu/transport/virtual/base.py:466]: if that improves the codebase you could give it a try

**Comment by auvipy** [comment]
also can you and your team please check the two open PR in pyamqp with this two PR in kombu & celery that nothing create any regression? I like your efforts very much, just being too cautious as it's holiday season coming.

File: kombu/transport/virtual/base.py (hunk 1 above)
Comment (pawl) [reply kombu/transport/virtual/base.py]: Ok, I went ahead and made this change: https://github.com/celery/kombu/pull/1470/commits/a439adaceae6c7e56e0d7c192ce63f2434a079e8

**Comment by auvipy** [comment]
@michael-lazar

**Review by auvipy — APPROVED** [review]

**Pull request merged**